From dce7a4000355c26718efe055ad6e1c54872ba656 Mon Sep 17 00:00:00 2001
From: Michael W <baddaemon87@gmail.com>
Date: Sat, 13 Jan 2018 16:32:23 +0100
Subject: [PATCH 066/249] wiki: Rework architecture to allow different archs
 for cpu and os

* Because 64 bit cpu can also run 32 bit os in some cases

Change-Id: I12e79666bf0f8da671479147465c1c11a6d7fc09
---
 _data/devices/Z008.yml                |  2 +-
 _data/devices/Z00A.yml                |  2 +-
 _data/devices/fugu.yml                |  2 +-
 _includes/deviceinfo.html             | 11 ++++++++---
 _includes/templates/device_install.md | 12 ++++++++++--
 device_sample/sample.yml              |  2 +-
 test/schema-06.yml                    | 25 +++++++++++++++++++++----
 7 files changed, 43 insertions(+), 13 deletions(-)

diff --git a/_data/devices/Z008.yml b/_data/devices/Z008.yml
index 311488e..6f71505 100644
--- a/_data/devices/Z008.yml
+++ b/_data/devices/Z008.yml
@@ -1,4 +1,4 @@
-architecture: x86_64
+architecture: {cpu: 'x86_64', userspace: 'x86'}
 battery: 3000mAh
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
diff --git a/_data/devices/Z00A.yml b/_data/devices/Z00A.yml
index 33a6c22..bd06255 100644
--- a/_data/devices/Z00A.yml
+++ b/_data/devices/Z00A.yml
@@ -1,4 +1,4 @@
-architecture: x86_64
+architecture: {cpu: 'x86_64', userspace: 'x86'}
 battery: 3000mAh
 bluetooth: {spec: '4.0', profiles: [A2DP]}
 cameras:
diff --git a/_data/devices/fugu.yml b/_data/devices/fugu.yml
index b4d6c71..8ce1531 100644
--- a/_data/devices/fugu.yml
+++ b/_data/devices/fugu.yml
@@ -1,4 +1,4 @@
-architecture: x86_64
+architecture: {cpu: 'x86_64', userspace: 'x86'}
 battery:
 bluetooth: {spec: '4.1'}
 cameras: []
diff --git a/_includes/deviceinfo.html b/_includes/deviceinfo.html
index 31c3824..1322a7f 100644
--- a/_includes/deviceinfo.html
+++ b/_includes/deviceinfo.html
@@ -80,12 +80,17 @@ ul, ol {
             {% endcase %} {{ device.cpu }}<br>{{ device.cpu_freq }}
         </td>
     </tr>
-    {% if device.architecture %}
     <tr>
         <td class="header">Architecture</td>
-        <td>{{ device.architecture }}</td>
+        <td>
+            {%- if device.architecture.cpu -%}
+            CPU: {{ device.architecture.cpu }}<br/>
+            Android: {{ device.architecture.userspace }}
+            {%- else -%}
+            {{ device.architecture }}
+            {%- endif -%}
+        </td>
     </tr>
-    {% endif %}
     <tr>
         <td class="header">GPU</td>
         <td>{{ device.gpu }}</td>
diff --git a/_includes/templates/device_install.md b/_includes/templates/device_install.md
index 169e930..0b890f0 100644
--- a/_includes/templates/device_install.md
+++ b/_includes/templates/device_install.md
@@ -27,12 +27,20 @@ There are no recovery install instructions for this discontinued device.
 
 ## Installing LineageOS from recovery
 
+{%- capture userspace_architecture -%}
+{%- if device.architecture.userspace -%}
+{{ device.architecture.userspace }}
+{%- else -%}
+{{ device.architecture }}
+{%- endif -%}
+{%- endcapture -%}
+
 {% if device.maintainers %}
 1. Download the [LineageOS install package](https://download.lineageos.org/{{ device.codename }}) that you'd like to install or [build]({{ "devices/" | append: device.codename | append: "/build" | relative_url }}) the package yourself.
 {% else %}
 1. [Build]({{ "devices/" | append: device.codename | append: "/build" | relative_url }}) a LineageOS install package.
 {% endif %}
-    * Optionally, download 3rd party application packages such as [Google Apps]({{ "gapps.html" | relative_url }})
+    * Optionally, download 3rd party application packages such as [Google Apps]({{ "gapps.html" | relative_url }}) (use the `{{ userspace_architecture }}` architecture)
 2. Place the LineageOS `.zip` package, as well as any other .zip packages on the root of `/sdcard`:
     * Using adb: `adb push filename.zip /sdcard/`
     * You can use any method you are comfortable with. `adb` is universal across all devices, and works both in Android and recovery mode, providing
@@ -47,7 +55,7 @@ There are no recovery install instructions for this discontinued device.
 9. Follow the on-screen prompts to install the package.
 10. _(Optional)_: Install any additional packages using the same method.
     {% include note.html content="If you want any Google Apps on your device, you must follow this step **before** the first reboot!" %}
-11. _(Optional)_: Root the device by installing the [LineageOS su add-on](https://download.lineageos.org/extras){% if device.architecture %} (use the `{{ device.architecture }}` package){% endif %} or using any other method you prefer.
+11. _(Optional)_: Root the device by installing the [LineageOS su add-on](https://download.lineageos.org/extras) (use the `{{ userspace_architecture }}` package) or using any other method you prefer.
 12. Once installation has finished, return to the main menu, select **Reboot**, and then **System**.
 
 ## Get assistance
diff --git a/device_sample/sample.yml b/device_sample/sample.yml
index 5e2977b..4891eaf 100644
--- a/device_sample/sample.yml
+++ b/device_sample/sample.yml
@@ -1,4 +1,4 @@
-architecture: arm
+architecture: {cpu: 'arm64', userspace: 'arm'}
 battery: Removable Li-Ion 2800 mAh
 bluetooth: {spec: 4.0 + LE}
 cameras:
diff --git a/test/schema-06.yml b/test/schema-06.yml
index b268142..ee7fb19 100644
--- a/test/schema-06.yml
+++ b/test/schema-06.yml
@@ -20,15 +20,32 @@ definitions:
     - 14.1
     - 15.1
     type: number
-properties:
-  architecture:
-    "$id": "/properties/architecture"
-    description: The CPU architecture of the device.
+  valid_architectures:
     enum:
     - arm
     - arm64
     - x86
     - x86_64
+    type: string
+  architecture_data:
+    properties:
+      cpus:
+        "$ref": "#/definitions/valid_architectures"
+      userspace:
+        "$ref": "#/definitions/valid_architectures"
+    required:
+    - cpu
+    - userspace
+    type: object
+properties:
+  architecture:
+    "$id": "/properties/architecture"
+    description: The CPU architecture of the device.
+    oneOf:
+    - description: A string defining the architecture of cpu AND userspace
+      "$ref": "#/definitions/valid_architectures"
+    - description: An object to provide different architectures for cpu and userspace
+      "$ref": "#/definitions/architecture_data"
   bluetooth:
     "$id": "/properties/bluetooth"
     properties:
-- 
2.17.1

